File: core/src/vct/col/ast/ClassType.java
package vct.col.ast;

import static hre.System.Abort;
import static hre.System.Debug;
import static hre.System.Warning;

import java.util.Arrays;

public class ClassType extends Type {

  @Override
  public <R,A> R accept_simple(ASTMapping1<R,A> map,A arg){
    return map.map(this,arg);
  }

  public static final ClassType label_type=new ClassType("<<label>>");
  public static final ClassType null_type=new ClassType("<<null>>");
      
  private final String name[];
  
  private ASTDeclaration def;
  
  public void setDefinition(ASTDeclaration def){
    this.def=def;
  }
  
  public ASTDeclaration getDefinition(){
    return def;
  }
  
  public ClassType(String name[],ASTNode ... args){
    super(args);
    if (name.length==0) Abort("illegal name of length 0");
    this.name=Arrays.copyOf(name,name.length);
  }
  
  public String[] getNameFull(){
    return Arrays.copyOf(name,name.length);
  }
  
  public ClassType(String name,ASTNode ... args) {
    super(args);
    this.name=new String[1];
    this.name[0]=name;
  }

  public String getName(){
    return name[name.length-1];
  }
  public String getFullName(){
    String res=name[0];
    for(int i=1;i<name.length;i++){
      res+="."+name[i];
    }
    return res;
  }

  
  @Override
  public <T> void accept_simple(ASTVisitor<T> visitor){
    try {
      visitor.visit(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
     }
      throw t;
    }
  }
  
  @Override
  public <T> T accept_simple(ASTMapping<T> map){
    try {
      return map.map(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
      }
      throw t;
    }
  }
 
  @Override
  public <T> T accept_simple(TypeMapping<T> map){
    try {
      return map.map(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
     }
      throw t;
    }
  }

  public boolean supertypeof(ProgramUnit context, Type t){
    // java.lang.Object is supertype of everything.
    if (name.length==3 && name[0].equals("java") &&
        name[1].equals("lang") && name[2].equals("Object")) return true;
    if (t instanceof ClassType) {
      ClassType ct=(ClassType)t;
      // no longer needed?? if (name.length==1 && name[0].equals("Object")) return true;
      // Everything is a supertype of <<null>>.
      if (ct.name.length==1 && ct.name[0].equals("<<null>>")) return true;
      if (ct.name.length==1 && ct.name[0].equals("<<label>>")) return true;
      return supertype_search(context,ct);
    }
    return false;
  }
  
  
  private boolean supertype_search(ProgramUnit context, ClassType ct) {
    Debug("checking if %s is asuper type of %s",this,ct);
    if (ct.name.length==name.length){
      int i=0;
      while(i<name.length){
        if (!ct.name[i].equals(name[i])) break;
        i++;
      }
      if (i==name.length) return true;
    }
    Debug("argument is not the same type not the same type");
    if (context==null) {
      Debug("missing context");
      return false;
    }
    ASTClass cl=context.find(ct);
    if (cl==null) {
      Debug("class not found");
      return false;
    }
    for(ClassType ct_parent:cl.super_classes){
      if (supertype_search(context,ct_parent)) return true;
    }
    return false;
  }

  public String toString(){
    String res=getFullName();
    if (this.args!=null && this.args.length>0){
      res=res+"<"+args[0];
      for(int i=1;i<args.length;i++) res=res+","+args[i];
      res=res+">";
    }
    return res;
  }
  
  public int hashCode(){
    return getFullName().hashCode();
  }
  public boolean equals(Object o){
    if (o instanceof ClassType) {
      return getFullName().equals(((ClassType)o).getFullName());
    } else {
      return false;
    }
  }
  public ASTNode zero(){
    return new NameExpression(ASTReserved.Null);
  }
  
  public boolean isNull(){
    return name.length==1 && name[0].equals("<<null>>"); 
  }
}

File: vercors/src/main/java/vct/col/ast/type/ClassType.scala
package vct.col.ast.`type`

import hre.lang.System.Debug
import vct.col.ast._
import vct.col.ast.expr.NameExpression
import vct.col.ast.generic.ASTNode
import vct.col.ast.stmt.decl.{ASTClass, ASTDeclaration, ProgramUnit}
import vct.col.ast.util.{ASTVisitor, TypeMapping}
import vct.col.util.{ASTMapping, ASTMapping1, VisitorHelper}

import scala.collection.JavaConverters._

object ClassType {
  val nullType = new ClassType("<<null>>")
  val labelType = new ClassType("<<label>>")
  
  /** Tests whether `name` holds the class type `java.lang.Object`. */
  def isJavaLangObject(name:List[String]) : Boolean =
    name == List("java","lang","Object") || name == List("java_DOT_lang_DOT_Object")
  
  /** Tests whether `name` holds a default type name, for example "null" or "label". */
  def isDefaultName(name:List[String]) : Boolean = 
    name == nullType.names || name == labelType.names
}

/**
 * AST node that represents the type of classes (including their class parameters).
 * 
 * @param names A list of name parts that together constitute the full class name (including package name)
 * @param params A list of AST nodes representing the types of the class parameters
 * @author sccblom, whmoortwijn
 */
case class ClassType(val names:List[String], val params:List[ASTNode]) extends Type(params) with VisitorHelper {
  require(!names.isEmpty, "class types must have a name (at least one name part).")
  
  /** Constructs a new class type from Java constructs. */
  def this(names:List[String], params:java.util.List[ASTNode]) = this(names, params.asScala.toList)
  
  /** Constructs a new class type from Java constructs. */
  def this(names:Array[String], params:java.util.List[ASTNode]) = this(names.toList, params)

  def this(names:Array[String], args:Array[ASTNode]) = this(names.toList, args.toList)
  def this(names:Array[String]) = this(names, Array[ASTNode]())
  def this(name:String) = this(Array(name))
  
  var definition : ASTDeclaration = null
  
  def getName = names.last
  def getNameFull = names.toArray
  def getFullName(separator:String) = names mkString separator
  def getFullName : String = getFullName(".")
  def setDefinition(decl:ASTDeclaration) = definition = decl

  /**
   * Checks if any of `foundClass`'s super classes (of implemented classes) is a supertype of `this`,
   * in the given program context (`context`).
   */
  private def searchContextForSupertype(context:ProgramUnit, foundClass:Option[ASTClass]) = foundClass match {
    case None => Debug("class not found."); false
    case Some(cl:ASTClass) => {
      cl.super_classes.exists(parent => searchForSupertype(Some(context), parent)) ||
      cl.implemented_classes.exists(parent => searchForSupertype(Some(context), parent))
    }
  }
  
  /**
   * Checks if the type of this object (i.e. `this`) is a supertype of `ct` in the 
   * given program context (`unit`).
   */
  private def searchForSupertype(unit:Option[ProgramUnit], ct:ClassType) : Boolean = {
    Debug(s"checking if $this is a supertype of $ct.")
    if (ct.names == this.names) true else unit match {
      case Some(context:ProgramUnit) => searchContextForSupertype(context, Option(context.find(ct)))
      case None => Debug("missing program context."); false
    }
  }
  
  /**
   * @note Everything is a supertype of `null` (or any other default name).
   */
  private def searchForSupertype(context:ProgramUnit, other:Type) : Boolean = other match {
    case ct:ClassType => ClassType.isDefaultName(ct.names) || searchForSupertype(Option(context), ct)
    case _ => false
  }
  
  /**
   * Tests whether this class is a supertype of `otherType` (in the given `context`).
   * 
   * @note `java.lang.Object` is a supertype of everything. However, implicit casts for anything other than class types
   *       is not supported (e.g. boxing, array conversion, ...) so we restrict it to class types for now, in order to
   *       generate a useful error.
   */
  override def supertypeof(context:ProgramUnit, otherType:Type) =
    (ClassType.isJavaLangObject(this.names) && otherType.isInstanceOf[ClassType]) ||
      searchForSupertype(context, otherType)
  
  override def equals(obj:Any) = obj match {
    case other:ClassType => this.getFullName == other.getFullName
    case _ => false
  }
  
  override def hashCode = this.getFullName.hashCode
  override def toString = this.getFullName + super.toString
  override def isNull = this.names == ClassType.nullType.names
  override def zero = new NameExpression(ASTReserved.Null)
  
  override def accept_simple[T,A](m:ASTMapping1[T,A], arg:A) = m.map(this, arg)
  override def accept_simple[T](v:ASTVisitor[T]) = handle_standard(() => v.visit(this))
  override def accept_simple[T](m:ASTMapping[T]) = handle_standard(() => m.map(this))
  override def accept_simple[T](m:TypeMapping[T]) = handle_standard(() => m.map(this))

  override def debugTreeChildrenFields: Iterable[String] = Seq("names", "args")
  override def debugTreePropertyFields: Iterable[String] = Seq()
}
